zk.js/src/utxo/utxo.ts

```diff
@@ -14,77 +14,134 @@ import {
+  owner: CompressionPublicKey | PublicKey;
+  /** Optional number of lamports and SPL amount assigned to the utxo */
   amounts: BN[];
+  /** Optional native mint and SPL mint address respective to 'amounts' */
   assets: PublicKey[];
-  assetsCircuit: string[];
-  blinding: BN;
+  /** Random value to force uniqueness of 'hash' */
+  blinding: BN31;
+  /** Optional type of utxo for custom circuits, defaults to 'native' */
+  type: string;
+  /** Default to '0' */
+  version: string;
+  /** Default to '0' */
   poolType: string;
-  utxoHash: string;
-  transactionVersion: string;
+  /** Indicator for whether the utxo is empty, for readability */
   isFillingUtxo: boolean;
-  utxoDataHash: BN;
-  utxoData?: any;
-  metaHash?: BN;
-  address?: BN;
+  /** Default 'true'. Whether the inputs to 'hash' are public or not. Useful for confidential compute. */
+  isPublic: boolean;
+  /** Optional persistent id of the utxo. Used for compressed PDAs and non-fungible tokens */
+  address?: BN31;
+  /** Optional public key of program that owns the metadata */
+  metadataOwner?: PublicKey;
+  /**
+   *	metadata which is immutable in normal transactions.
+   *	metadata can be updated by the metadataOwner with a dedicated system psp.
+   */
+  metadata?: any; /// TODO: add metadata type
+  // /** hash of metadata */
+  // metadataHash?: string;
+  /** hash of metadataHash and metadataOwner */
+  metaHash?: BN31;
+};
+
+/** Utxo that had previously been inserted into a state Merkle tree */
+export type Utxo = Omit<BaseUtxo, "owner"> & {
+  /** Compression public key of the user that owns the utxo */
+  owner: CompressionPublicKey;
+  /** Hash that invalidates utxo once inserted into nullifier queue, if isPublic = true it defaults to: 'hash' */
+  nullifier: BN31;
+  /** Numerical identifier of the Merkle tree which the 'hash' is part of */
+  merkletreeId: number;
+  /** Proof path attached to the utxo. Can be reconstructed using event history */
+  merkleProof: string[];
+  /** Index of 'hash' as inserted into the Merkle tree. Max safe tree depth using number type would be **52, roughly 4.5 x 10^15 leaves */
+  merkleTreeLeafIndex: number;
+};
+
+/** Utxo that is not inserted into the state tree yet. */
+export type OutUtxo = Omit<BaseUtxo, "owner"> & {
+  /** Compression public key of the user that owns the utxo */
+  owner: CompressionPublicKey;
+  /**
+   * Optional public key of the ouput utxo owner once inserted into the state tree.
+   * Only set if the utxo should be encrypted asymetrically.
+   */
+  encryptionPublicKey?: Uint8Array;
+};
+
+/** Type safety: enforce that the utxo is not encrypted */
+export type Public = { isPublic: true };
+export type PublicBaseUtxo = BaseUtxo & Public;
+export type PublicUtxo = Utxo & Public;
+export type PublicOutUtxo = OutUtxo & Public;
+
+export type NullifierInputs = {
+  signature: BN;
+  /** hash of the utxo preimage */
+  hash: BN31;
```

Comment:
I like the idea of this type. To make it better I believe we should make call it BN254 and extend constructor to check that the number is smaller than FIELD_SIZE eg.
```
BN254 extends BN
constructor() {
super()
let input = new BN(input)
if (input.gte(FIELD_SIZE)
  throw Error()
}
```